lib/rules/no-dupe-disjunctions.ts

```diff
@@ -9,6 +10,492 @@ import type {
+): Iterable<Result> {
+    const { flags, toCharSet } = context
+
+    const isCoveredOptions: Parameters<typeof isCoveredNode>[2] = {
+        flags,
+        canOmitRight: hasNothingAfter,
+        toCharSet,
+    }
+    const isCoveredOptionsNoPrefix: Parameters<typeof isCoveredNode>[2] = {
+        flags,
+        canOmitRight: false,
+        toCharSet,
+    }
+
+    for (let i = 0; i < alternatives.length; i++) {
+        const alternative = alternatives[i]
+
+        for (let j = 0; j < i; j++) {
+            const other = alternatives[j]
+
+            if (isCoveredNode(other, alternative, isCoveredOptions)) {
+                if (isEqualNodes(other, alternative, toCharSet)) {
+                    yield {
+                        type: "Duplicate",
+                        alternative,
+                        others: [other],
+                    }
+                } else if (
+                    hasNothingAfter &&
+                    !isCoveredNode(other, alternative, isCoveredOptionsNoPrefix)
+                ) {
+                    yield {
+                        type: "PrefixSubset",
+                        alternative,
+                        others: [other],
+                    }
+                } else {
+                    yield { type: "Subset", alternative, others: [other] }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Tries to find duplication in the given alternatives.
+ *
+ * It will search for prefix duplications. I.e. the alternative `ab` in `a|ab`
+ * is a duplicate of `a` because if `ab` accepts, `a` will have already accepted
+ * the input string. This makes `ab` effectively useless.
+ *
+ * This operation will modify the given NFAs.
+ */
+function* findPrefixDuplicationNfa(
+    alternatives: [NFA, Alternative][],
+): Iterable<Result> {
+    if (alternatives.length === 0) {
+        return
+    }
+
+    // For two alternatives `A|B`, `B` is useless if `B` is a subset of `A[^]*`
+
+    const all = NFA.all({ maxCharacter: alternatives[0][0].maxCharacter })
+
+    for (let i = 1; i < alternatives.length; i++) {
+        const [nfa, alternative] = alternatives[i]
+
+        const overlapping = alternatives
+            .slice(0, i)
+            .filter(([otherNfa]) => !nfa.isDisjointWith(otherNfa))
+
+        if (overlapping.length >= 1) {
+            const othersNfa = unionAll(overlapping.map(([n]) => n))
+            const others = overlapping.map(([, a]) => a)
+
+            // Only checking for a subset relation is sufficient here.
+            // Duplicates are VERY unlikely. (Who would use alternatives
+            // like `a|a[^]*`?)
+
+            if (isSubsetOf(othersNfa, nfa)) {
+                yield { type: "PrefixSubset", alternative, others }
+            } else {
+                // The else case is interesting.
+                // It means that _some_ of the paths in the current alternative are useless.
+                // TODO: Decide whether this should be reported as well.
```

Comment:
What do we want to do here?

In this case, the alternative is partially covered by previous alternatives. Should we report this too? E.g. `\d|\w`, `Foo|\w`.

Reply:
I think `\d|\w` can delete `\d`, so we may report it. However, I don't think `Foo|\w` can remove one alt. I don't think this rule should be reported. I think it might be nice to have an option to check for partial duplication or another rule.

Reply:
Perhaps I misunderstood. Did you say that there is a partial duplication of `(?:\d|\w)|(?:Foo|\w)`?
I think it would be useful to be able to tell the user the correct `\w` location.

Reply:
> I think it might be nice to have an option to check for partial duplication or another rule.

Then let's add an option. Adding another rule for this would also be possible but we would have to do all the NFA operations again.

> Perhaps I misunderstood.

No, you didn't. :)

Reply:
I thought about this some more. `\d|\w` is an interesting case and will be reported.

However, `Foo|\w` will not be reported. I mean, the "fixed" pattern would be `Foo|F(?!oo)|[a-zA-EG-Z0-9_]`. There's no point in reporting this.